[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes

larryliu0820 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99560
* __->__ #95849

Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will handle returning `np.ndarray` and add support for ndarray methods

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

larryliu0820 committed:
[dynamo] Add support for Tensor.numpy(), torch.from_numpy() and np.meshgrid()
As titled. Added `NumpyTensorVariable` to represent numpy.ndarray, hookup the numpy handler with `numpy_pytorch_interop`.

[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo] Add support for Tensor.numpy(), torch.from_numpy() and np.meshgrid()"


Issue: #93684

As titled. Added `NumpyTensorVariable` to represent numpy.ndarray, hookup the numpy handler with `numpy_pytorch_interop`.

TODO: make the `TensorVariable` getting back from `from_numpy()` non-resizable.

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo] Add support for Tensor.numpy(), torch.from_numpy() and np.meshgrid()"


Issue: #93684

As titled. Added `NumpyTensorVariable` to represent numpy.ndarray, hookup the numpy handler with `numpy_pytorch_interop`.

TODO: make the `TensorVariable` getting back from `from_numpy()` non-resizable.

cc soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

larryliu0820 commented:
This breaks a few unit tests, because they were expecting a graph break before but now they don't. For example:

```python
def ref_take(src, idx):
            if dtype == torch.bfloat16:
                src = src.half()
            src = src.cpu().numpy()
            idx = idx.cpu().numpy()
            out = torch.from_numpy(np.take(src, idx)).to(device=device, dtype=dtype)
            return out
```
Before out is giving np.ndarray correctly but now it gives tensor. Seems like we need to reimplement `NumpyTensorVariable`?

yanboliang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -604,6 +622,46 @@ def inline_torch_function_unwrapped(
             return tx.inline_user_function_return(tf_func_var, tf_args, {})
 
 
+class NumpyTensorVariable(TensorVariable):
```
What is the equivalent Python type of this variable? Is ```numpy.ndarray```? If yes, then why it extends from ```TensorVariable```. We should have ```python_type```, ```as_python_constant``` and  ```reconstruct``` for variables to facilitate python - dynamo interoperation.

lezcano replied:
Doesn't this still apply?

ezyang replied:
Yeah, this still applies

larryliu0820 committed:
Update on "[dynamo] Add support for Tensor.numpy(), torch.from_numpy() and np.meshgrid()"


Issue: #93684

As titled. Added `NumpyTensorVariable` to represent numpy.ndarray, hookup the numpy handler with `numpy_pytorch_interop`.

TODO: make the `TensorVariable` getting back from `from_numpy()` non-resizable.


[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo] Add support for Tensor.numpy(), torch.from_numpy() and np.meshgrid()"


Issue: #93684

As titled. Added `NumpyTensorVariable` to represent numpy.ndarray, hookup the numpy handler with `numpy_pytorch_interop`.

TODO: make the `TensorVariable` getting back from `from_numpy()` non-resizable.


[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo] Add support for Tensor.numpy(), torch.from_numpy() and np.meshgrid()"


Issue: #93684

As titled. Added `NumpyTensorVariable` to represent numpy.ndarray, hookup the numpy handler with `numpy_pytorch_interop`.

TODO: make the `TensorVariable` getting back from `from_numpy()` non-resizable.


[ghstack-poisoned]

yanboliang commented on torch/_dynamo/config.py:
```diff
@@ -198,6 +198,9 @@
 base_dir = dirname(dirname(dirname(abspath(__file__))))
 
+# trace through numpy ndarray as tensor and try to translate numpy function to torch function.
+trace_numpy = False
```
Can we have a better config name? E.g, numpy_ndarray_as_tensor?

ezyang replied:
Also, NGL, let's figure out how to make this true by default.

yanboliang commented on torch/_dynamo/variables/builder.py:
```diff
@@ -932,7 +935,9 @@ def _clone_input(value):
             )
 
-    if isinstance(example_value, torch.Tensor):
+    if isinstance(example_value, torch.Tensor) or (
```
I'm a bit confused here, can you check the following lines?
https://github.com/pytorch/pytorch/blob/fe05266fda4f908130dea7cbac37e9264c0429a2/torch/_dynamo/variables/builder.py#L403-L412

If the variable is ```numpy.ndarray```, it has been wrapped as ```NumpyVariable```. I think we should gate with ```config.trace_numpy```. If it's true, we always use ```NumpyTensorVariable```, otherwise use ```NumpyVariable```.

larryliu0820 replied:
Hmm I think `NumpyVariable` is wrapping `numpy` module, where `NumpyTensorVariable` is for `numpy.ndarray`? Similar to `TorchVariable` and `TensorVariable`.

yanboliang replied:
Actually no, check these lines:
https://github.com/pytorch/pytorch/blob/fe05266fda4f908130dea7cbac37e9264c0429a2/torch/_dynamo/allowed_functions.py#L277-L281

ezyang commented on torch/_dynamo/variables/misc.py:
```diff
@@ -860,7 +860,35 @@ def __init__(self, value, **kwargs):
+                meshgrid = torch_np._detail.implementations.meshgrid
+            else:
+                # should we give a warning since torch.meshgrid is having a different semantics than np?
+                meshgrid = torch.meshgrid
```
No, you should bail out with unimplemented("numpy") instead here

ezyang commented on torch/_dynamo/variables/torch.py:
```diff
@@ -289,6 +289,29 @@ def call_function(
             )
+        elif self.value is torch.from_numpy:
+            if not config.trace_numpy:
+                unimplemented("numpy")
```
It would be cool to suggest trace_numpy in the graph break message here

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -337,7 +337,27 @@ def call_method(
-        elif name in ("tolist", "numpy", "backward", "data_ptr"):
+        elif name == "numpy":
+            if not config.trace_numpy:
+                unimplemented(f"Tensor.{name}")
```
ditto here; suggesting here is more important, probably

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -604,6 +624,52 @@ def inline_torch_function_unwrapped(
+
+    def var_getattr(self, tx, name):
+        try:
+            super().var_getattr(tx, name)
```
This is wrong, don't do a blanket passthrough here

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -604,6 +624,52 @@ def inline_torch_function_unwrapped(
+        shape=None,
+        **kwargs,
+    ):
+        self.shape = shape
```
It's weird that you have to track this separately 🤔

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -604,6 +624,52 @@ def inline_torch_function_unwrapped(
+            unimplemented(f"numpy_ndarray.{name}")
+
+    def call_isinstance(self, tensor_type):
+        return False
```
This also looks wrong

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -604,6 +624,52 @@ def inline_torch_function_unwrapped(
+        args: "List[VariableTracker]",
+        kwargs: "Dict[str, VariableTracker]",
+    ) -> "VariableTracker":
+        unimplemented(f"numpy_ndarray.{name}")
```
Want to use the mappings in https://github.com/Quansight-Labs/numpy_pytorch_interop/blob/main/torch_np/_ndarray.py

ezyang commented on torch/_dynamo/variables/misc.py:
```diff
@@ -860,7 +860,35 @@ def __init__(self, value, **kwargs):
+        from .tensor import NumpyTensorVariable
+
+        options = VariableTracker.propagate([[self]], [args], [list(kwargs.values())])
+        if self.value is np.meshgrid:
```
Yeah, just implement the generic thing in `torch_np`. Get out the name of the Numpy function, and then do a getattr to get the corresponding thing from torch_np. You'll get a lot of stuff for free that way.

ezyang commented on test/dynamo/test_functions.py:
```diff
@@ -819,6 +823,24 @@ def augment(x: torch.Tensor) -> torch.Tensor:
 
+    @make_test
+    @requires_numpy_pytorch_interop
+    def test_numpy_meshgrid(x, y):
```
wait, how did we generate these input tensors 🤔

larryliu0820 replied:
Maybe [here](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/testing.py#L213)

ezyang commented:
Looks promising! More to do. Try it on some E2E examples!

larryliu0820 committed:
Update on "[dynamo] Add support for Tensor.numpy(), torch.from_numpy() and np.meshgrid()"


Issue: #93684

As titled. Added `NumpyTensorVariable` to represent numpy.ndarray, hookup the numpy handler with `numpy_pytorch_interop`.

TODO: make the `TensorVariable` getting back from `from_numpy()` non-resizable.


[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo] Add support for Tensor.numpy(), torch.from_numpy() and np.meshgrid()"


Issue: #93684

As titled. Added `NumpyTensorVariable` to represent numpy.ndarray, hookup the numpy handler with `numpy_pytorch_interop`.

TODO: make the `TensorVariable` getting back from `from_numpy()` non-resizable.


[ghstack-poisoned]

ezyang commented:
please retag me when you want more review

larryliu0820 committed:
Update on "[dynamo] Add support for Tensor.numpy(), torch.from_numpy() and np.meshgrid()"


Issue: #93684

As titled. Added `NumpyTensorVariable` to represent numpy.ndarray, hookup the numpy handler with `numpy_pytorch_interop`.

TODO: make the `TensorVariable` getting back from `from_numpy()` non-resizable.


[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo] Add support for Tensor.numpy(), torch.from_numpy() and np.meshgrid()"


Issue: #93684

As titled. Added `NumpyTensorVariable` to represent numpy.ndarray, hookup the numpy handler with `numpy_pytorch_interop`.

TODO: make the `TensorVariable` getting back from `from_numpy()` non-resizable.


[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo] Add support for Tensor.numpy(), torch.from_numpy() and np.meshgrid()"


Issue: #93684

As titled. Added `NumpyTensorVariable` to represent numpy.ndarray, hookup the numpy handler with `numpy_pytorch_interop`.

TODO: make the `TensorVariable` getting back from `from_numpy()` non-resizable.


[ghstack-poisoned]

ezyang commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -1941,6 +1941,26 @@ def RETURN_VALUE(self, inst):
+        # This is a hack to handle NumpyTensorVariable before returning it. This actually calls Tensor.numpy() to
+        # convert tensor to ndarray. Note that this doesn't handle TupleVariable so if more than one return values
+        # and one of them is NumpyTensorVariable this doesn't convert it and the result will be tensor instead of
+        # ndarray.
```
Why can't you do this properly with the reconstruct in the VariableTracker?

yanboliang replied:
@ezyang I think we should convert the fake_numpy_tensor back to numpy at ```RETURN_VALUE```. Look at this example:
```
def fn(x, y):
    a = x.numpy()
    b = y.numpy()
    torch._dynamo.graph_break()
    a = numpy.sin(a + 1)
    return a.T + b

x = torch.rand(2, 3)
y = torch.rand(3, 2)
print(fn(x, y))
```
If we do nothing, the return value is ```NumpyTensorVariable``` which is actually a tensor when switch back to python world. My suggestion here is to wrap these logics as a ```NumpyTensorVariable``` member method for better maintenance. 

> Why can't you do this properly with the reconstruct in the VariableTracker?

Do you mean to support graph break between many numpy ops(sth like in my example)? If yes, we should wrap the fake_numpy_tensor, compile the partial graph and get the numpy value when handling the graph break. Then during reconstruction, we directly load the numpy value and wrap back to  fake_numpy_tensor. Let me know if we are on the same page. Thanks.

ezyang replied:
we talked about this briefly; while we do need logic in reconstruct, we need different logic for RETURN_VALUE, and Yanbo will figure out where it should go

larryliu0820 replied:
Moved to output_graph.py as well as tensor.py. Need to handle tuple variable later.

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -604,6 +627,121 @@ def inline_torch_function_unwrapped(
+            result = TupleVariable(sizes)
+        elif name == "ndim" and self.ndim is not None:
+            result = ConstantVariable(self.ndim, **options)
+        elif name == "T":
```
Having to manually handle the attribute accesses here seems suboptimal; it would have been better if torch_np handled it. If possible, I'd prefer to land the PR *without* this code.

lezcano replied:
This is handled in `torch_np`, yes. These functions dispatch to the correct `Tensor` functions.

yanboliang replied:
Since ```NumpyTensorVariable``` extends from ```TensorVariable```, why don't just call the parent's ```var_getattr```?

ezyang commented on torch/_dynamo/variables/misc.py:
```diff
@@ -860,7 +860,32 @@ def __init__(self, value, **kwargs):
+
+        options = VariableTracker.propagate([[self]], [args], [list(kwargs.values())])
+        # lookup method name in torch_np
+        if self.value.__name__ in torch_np._detail.implementations.__dict__:
```
@rgommers is this the correct way to look up the implementation of a function from torch_np? This... doesn't seem quite right to me, IMO.

larryliu0820 replied:
hmm would it be better if I change it to inspect.getattr_static

ezyang replied:
So, I guess what I was expecting, is either `torch_np` has some mapping of name to function, so that you just do a lookup from the mapping, or there is some convention that `torch_np.XXX` corresponds to `numpy.XXX`, so you do a getattr on the module itself (with possibly some handling for numpy.submodule)

rgommers replied:
> or there is some convention that `torch_np.XXX` corresponds to `numpy.XXX`,

that should be the case

larryliu0820 replied:
So I can't just use `torch_np.meshgrid` since it's mapping numpy.meshgrid return (a list of ndarray)

larryliu0820 replied:
using `torch_np._details` instead

lezcano replied:
> So I can't just use torch_np.meshgrid since it's mapping numpy.meshgrid return (a list of ndarray)

Can you elaborate on that? The `torch_np` implementations and the `torch_np.ndarray` object are designed to be drop-in replacements for `np` functions. You should be able to use them here (in fact, that's their only reason of existence). Even more, you should not use things from the repo with an underscore, as those are meant to be implementation details.

larryliu0820 replied:
@lezcano: `torch_np.meshgrid` currently takes in 2 torch tensors but returns a list of np ndarrays. Directly using that doesn't fit what we want - we want torch tensor in torch tensor out. Instead `torch_np._detail.implementations.meshgrid` is what we want.

lezcano replied:
That's not quite right. All the functions in `torch_np` accept `torch_np.ndarray`s (and other array-like things that NumPy accepts like lists of lists or simply python or numpy scalars) and return a `torch_np.ndarray`. `torch_np.ndarray` is just a thin wrapper around `torch.Tensor`, so here, you should simply access `ndarray.tensor` and use that internally.

larryliu0820 replied:
@lezcano hmm it's not simple as that, I understand we need to pass `torch_np.ndarray` around for `torch_np.meshgrid` and other APIs, the thing is on dynamo (AFAIK) we can't support arbitrary class, even it's just a simple wrapper of torch.Tensor. 

My understanding for how dynamo works is that it when we call `np.meshgrid`, we will need to translate that to some other function, but only taking torch.Tensor as input and output. Now if I use `torch_np.meshgrid` I can't just get back the result and access `ndarray.tensor` in `NumpyTensorVariable`, since we don't actually run `torch_np.meshgrid` during dynamo compilation and hence no result will be given.

With that said I can try and see if I can write a simple local wrapper function for `torch_np.meshgrid` to get the Tensor out. Correct me if I'm wrong @ezyang @yanboliang

lezcano replied:
The way we designed `torch_np`, and how I see this is as follows:
- `torch_np` provides a class `torch_np.ndarray` that can be created from an `np.ndarray`
  - Internally, this class is a wrapper around a `torch.Tensor` that effectively implements the API from `np.ndarray` via `torch.` functions and `torch.Tensor` methods.
- The other thing that `torch_np` is implementations for all the free functions (and a few other funny objects) from the NumPy API in terms of free functions within the `torch` namespace.
- All these functions and methods accept anything that NumPy accepts (`np.ndarray`s, Python scalars, NumPy scalars, list of lists... anything that smells like a `np.ndarray` really), np.dtypes, and other NumPy-lang objects, wraps these into `torch_np.ndarray`, `torch.dtype`s and equiv. Torch-lang objects, calls one or more PyTorch operations, and returns one (or more) `torch_np.ndarray`s.

To summarise, `torch_np` functions are functions from NumPy-objects + `torch_np.ndarray`s into `torch_np.ndarray`s implemented via PyTorch functions. These functions return `torch_np.ndarray`s so that we can compose calls to their methods.

With this tooling, the way I figured this would be implemented is:
- If we see a call to an `np.foo`, dispatch it to `torch_np.foo`. Don't worry about wrapping the `np.arrays` or `np.dtypes` here, `torch_np` functions will take care fo that.
- If we see a call to an `np.ndarray` method or property or anything, wrap the object it's being called on via `torch_np.asarray` and call the same method/property on the resulting `torch_np.ndarray`.
  - After doing this, if you keep tracing, you know that all you're going to see here is good old calls to functions and methods within the torch namespace.
- Finally, when you are leaving the frame, you'll still need to convert back the `torch_np.ndarray`s into `np.array`s but that should be it.

This being said, let's see what Ed and Yanbo have to say.

ezyang replied:
Yeah, storing a torch_np.ndarray inside the NumpyVariable seems nice and convenient. I think the proxy stuff will work out too?

larryliu0820 replied:
OK I did some plumbing and it seems to be working. Now `NumpyTensorVariable` is storing a `torch_np.ndarray`. The only issue I'm having is there's no way to `LOAD_ATTR` if I need to return an actual numpy ndarray.

For example:

```
def f(x: torch.Tensor):
    a = x.numpy()
    return a.T
```

Needs to be translated into something like:
```
def f(x: torch.Tensor):
    b = __compiled_function(x)
    return b.tensor.numpy()
```
Currently I'm adding some instructions for `.tensor` and `.numpy()` but for `.tensor` it failed because `tensor` doesn't exist in `code.co_names`. The workaround I can think of, is to add a method API `.to_numpy()` to `torch_np.ndarray`. Thoughts?

larryliu0820 replied:
> there's no way to LOAD_ATTR if I need to return an actual numpy ndarray

This is resolved. Thanks to @yanboliang

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch Tensor to back a `VariableTracker` for ndarray. Say a new `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch tensor equivalent.
* Leverage numpy & torch interop library to save some manual work.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`
4. returning a numpy ndarray instead of a tensor (this is done by adding some instructions at the end of `RETURN_VALUE()`)

Next PR will add support for ndarray methods

[ghstack-poisoned]

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -585,6 +585,16 @@ def compile_subgraph(
+        # This is a hack to handle NumpyTensorVariable before returning it. This actually calls Tensor.numpy() to
+        # convert tensor to ndarray. Note that this doesn't handle TupleVariable so if more than one return values
+        # and one of them is NumpyTensorVariable this doesn't convert it and the result will be tensor instead of
+        # ndarray.
```
Uhh, so did we not agree how to fix this? Is this going to go away before merging?

larryliu0820 replied:
Are you referring to the comment? If so I can make it better. I think we agreed on the approach, which is to add special logic to handle `NumpyTensorVariable` in `compile_subgraph()`.

larryliu0820 replied:
I'll update this

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch Tensor to back a `VariableTracker` for ndarray. Say a new `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch tensor equivalent.
* Leverage numpy & torch interop library to save some manual work.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`
4. returning a numpy ndarray instead of a tensor (this is done by adding some instructions at the end of `RETURN_VALUE()`)

Next PR will add support for ndarray methods

[ghstack-poisoned]

lezcano commented on torch/_dynamo/allowed_functions.py:
```diff
@@ -225,7 +225,7 @@ def _builtin_function_ids():
     rv = dict()
     if HAS_NUMPY:
-        for mod in (np, np.random):
+        for mod in (np, np.random, np.meshgrid):
```
What's `meshgrid`doing here? Is this a leftover?

ezyang commented on torch/_dynamo/output_graph.py:
```diff
@@ -585,6 +585,12 @@ def compile_subgraph(
+        if stack_values and isinstance(stack_values[-1], NumpyTensorVariable):
+            self.add_output_instructions(
+                stack_values[-1].reconstruct_ndarray(stack_values, tx)
+            )
```
So what's the deal? Why do we still have to do this?

larryliu0820 replied:
Yeah trying to figure this out. Let me remove this thing and add `reconstruct` support in next PR.

lezcano commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -337,7 +343,31 @@ def call_method(
+                unimplemented(
+                    f"Tensor.{name}. Turn on config.numpy_ndarray_as_tensor to support tensor.numpy()."
+                )
+            from torch_np import array
```
You probably just want to feed it to the constructor of `torch_np.ndarray`.

larryliu0820 commented on torch/_dynamo/variables/torch.py:
```diff
@@ -289,6 +291,32 @@ def call_function(
+            t = args[0]
+            from .tensor import NumpyTensorVariable
+
+            if isinstance(t, NumpyTensorVariable):
+                # TODO: mark the tensor as non-resizable
+                return wrap_fx_proxy_cls(
+                    target_cls=TensorVariable,
+                    tx=tx,
+                    proxy=tx.output.create_proxy(
+                        "call_function",
+                        torch_np._helpers.ndarrays_to_tensors,
+                        *proxy_args_kwargs(args, {}),
+                    ),
+                    example_value=None,
+                    **options,
+                )
```
That's a problem because in Python bytecode, `LOAD_ATTR` needs to load the name from `co_names` in code. Since we don't have `tensor` in `co_names`, `arr.tensor` won't work.

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch Tensor to back a `VariableTracker` for ndarray. Say a new `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch tensor equivalent.
* Leverage numpy & torch interop library to save some manual work.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`
4. returning a numpy ndarray instead of a tensor (this is done by adding some instructions at the end of `RETURN_VALUE()`)

Next PR will add support for ndarray methods

[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will add support for ndarray methods

[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will add support for ndarray methods

[ghstack-poisoned]

larryliu0820 commented:
@pytorchbot rebase

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will handle returning `np.ndarray` and add support for ndarray methods

[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will handle returning `np.ndarray` and add support for ndarray methods

[ghstack-poisoned]

lezcano commented:
@thomasjpfan did some experiments with this PR and the compat layer, and it seems like there are still quite a few things that should be sorted. See [this notebook](https://gist.github.com/thomasjpfan/513115f8c6265b83c9fe69ec9f02f11a).

ezyang commented on torch/_dynamo/variables/misc.py:
```diff
@@ -559,7 +564,30 @@ def __init__(self, value, **kwargs):
+        from .builder import wrap_fx_proxy_cls
+        from .tensor import NumpyTensorVariable
+
+        options = VariableTracker.propagate([[self]], [args], [list(kwargs.values())])
```
this is not the idiomatic way to do it (although I forget what the idiomatic way is)

ezyang commented:
This seems nearly ready to go, just need to resolve the TensorVariable subclassing situation. What might be the right thing to do is have NumpyNdarrayVariable CONTAIN a TensorVariable, which may make it more convenient to forward operations. Or maybe not, have to see the code both ways.

larryliu0820 commented:
> @thomasjpfan did some experiments with this PR and the compat layer, and it seems like there are still quite a few things that should be sorted. See [this notebook](https://gist.github.com/thomasjpfan/513115f8c6265b83c9fe69ec9f02f11a).

Thanks for the notebook, let me try to address them in a separate PR. This PR is setting the foundation and we can iterate on it once committed.

lezcano commented:
Note that that notebook shows how `np.mean(X, 1)` fails on this PR as it seems to try to trace through NumPy rather than PyTorch. I think this PR should at least fix this point, and then add a more thorough testing before being merged.

larryliu0820 commented:
> Note that that notebook shows how `np.mean(X, 1)` fails on this PR as it seems to try to trace through NumPy rather than PyTorch. I think this PR should at least fix this point, and then add a more thorough testing before being merged.

Check out this test:
```python
    @make_test
    @requires_numpy_pytorch_interop
    def test_mean_sum_np(x: torch.Tensor):
        x_mean = np.mean(x.numpy(), 1)
        x_sum = np.sum(x_mean)  # x_sum is a scalar, torch_np.sum(x_mean) is a ndarray, need to unify
        x_sum_array = np.asarray(x_sum)
        return torch.from_numpy(x_sum_array)
```

Basically this PR only supports torch.Tensor in torch.Tensor out. Also notice that `np.sum()` gives a scalar and `torch_np.sum()` gives a `torch_np.ndarray` so we need to unify the type, by calling `np.asarray()`

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will handle returning `np.ndarray` and add support for ndarray methods

[ghstack-poisoned]

larryliu0820 commented:
> This seems nearly ready to go, just need to resolve the TensorVariable subclassing situation. What might be the right thing to do is have NumpyNdarrayVariable CONTAIN a TensorVariable, which may make it more convenient to forward operations. Or maybe not, have to see the code both ways.

So it seems we don't even need to contain a TensorVariable? I tested on the current version and it works.

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will handle returning `np.ndarray` and add support for ndarray methods

[ghstack-poisoned]

ezyang reviewed (approved):
approving to move things along

larryliu0820 commented:
@pytorchbot rebase

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will handle returning `np.ndarray` and add support for ndarray methods

[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will handle returning `np.ndarray` and add support for ndarray methods

[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will handle returning `np.ndarray` and add support for ndarray methods

[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will handle returning `np.ndarray` and add support for ndarray methods

[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will handle returning `np.ndarray` and add support for ndarray methods

[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will handle returning `np.ndarray` and add support for ndarray methods

[ghstack-poisoned]

larryliu0820 committed:
Update on "[dynamo][numpy] Add NumpyTensorVariable to translate ndarray attribute calls to tensor attributes"


Issue: #93684

# Problem

Reduce graph breaks when dynamo compiles python functions containing numpy functions and ndarray operations.

# Design (as I know it)

* Use torch_np.ndarray(a wrapper of tensor) to back a `VariableTracker`: `NumpyTensorVariable`.
* Translate all attributes and methods calls, on ndarray, to torch_np.ndarray equivalent.

This PR adds `NumpyTensorVariable` and supports:
1.  tensor to ndarray, ndarray to tensor
2. numpy functions such as numpy.meshgrid()
3. ndarray attributes such as `itemsize`, `stride`

Next PR will handle returning `np.ndarray` and add support for ndarray methods

[ghstack-poisoned]
